### PR TITLE
Handle xarray DataArray in wrapped ufuncs

### LINKED ISSUE #66: Plans for full xarray compatibility?
The toolbox is broadly compatible with [xarray](https://github.com/pydata/xarray) DataArrays, but struggles when needing to do things such as broadcast along specific dimensions (a particularly powerful feature of xarray objects), or when xarray objects are separated into "chunks" using [dask](https://github.com/dask/dask). The broadcasting issue can be worked around using the xarray `apply_ufunc` command, but this requires writing a wrapper function for each `gsw` function. That approach still fails to handle chunked arrays, so that the workaround is to extract the underlying numpy arrays from the DataArrays and then plug the output of the `gsw` function back into a new DataArray. Some example wrapper functions can be seen [here](https://github.com/gmacgilchrist/wmt_bgc/blob/master/wmt_bgc/basic.py).

I'm wondering if there are any plans to improve this compatibility with xarray, essentially to allow xarray DataArrays (as well as numpy arrays) to be passed directly to the `gsw` functions, or whether I would be better placed to try to write an overall "wrapper package" to brute force compatibility.

### PR DESCRIPTION
**Pull request by efiring**
Closes #66 and #65, unless someone comes up with an example illustrating that there is more we need to do.

This is based on `__array_ufunc__`; it's not clear whether anything .

Tests include running all the check functions with minimal xarray DataArrays for SA_check_cast, t_check_cast, and p_check_cast. Additional tests include one based on the following example.

```python
import gsw
import xarray as xr
import numpy as np
import dask.array as dsa

# define some input data
shape = (100, 1000)
chunks = (100, 200)
sp = xr.DataArray(dsa.full(shape, 35., chunks=chunks), dims=['time', 'depth'])
p = xr.DataArray(np.arange(shape[1]), dims=['depth'])
lon = 0
lat = 45

sa = gsw.SA_from_SP(sp, p, lon, lat)
sa.compute()
```

### CODE REVIEW

File: gsw/_utilities.py [hunk 1]
```diff
@@ -26,20 +26,35 @@ def wrapper(*args, **kw):
             args = list(args)
             args.append(p)
 
-        isarray = np.any([hasattr(a, '__iter__') for a in args])
-        ismasked = np.any([np.ma.isMaskedArray(a) for a in args])
+        isarray = [hasattr(a, '__iter__') for a in args]
+        ismasked = [np.ma.isMaskedArray(a) for a in args]
+        isduck = [hasattr(a, '__array_ufunc__')
```
Comment (ocefpaf) [inline gsw/_utilities.py:31]: :smile: 

Maybe `isarrayinterface`? No string opinions, I actually like `isduck` ;-p

**Review by ocefpaf — APPROVED** [review]
Do you want to add tests or should we merge this for now?

**Comment by DocOtak** [comment]
Playing around by hand, everything seemed to magically work as expected, dask array chunks, special xarray broadcasting, even return object types (eg xr.DataSet) and coordinate propagation for xarray things. Basically, this is amazing and I think it what is actually wanted with respect to dask/xarray support.

That said, I've only just gotten started with the chunked/"bigish" data stuff myself, so I'd really want someone like @rabernat to take a look at this.

**Comment by efiring** [comment]
@ocefpaf, let's hold off on merging for a couple days; I would like to add at least some minimal tests.

**Comment by ocefpaf** [comment]
> Playing around by hand, everything seemed to magically work as expected, dask array chunks, special xarray broadcasting, even return object types (eg xr.DataSet) and coordinate propagation for xarray things. Basically, this is amazing and I think it what is actually wanted with respect to dask/xarray support.

Same here. All my "basic" tests worked. I guess we had almost everything we needed in the match_args already. Thanks @efiring!

**Comment by efiring** [comment]
I think this is good enough to merge, so I will proceed.  That should facilitate additional testing and recommendations for more work, if needed.

**Comment by ocefpaf** [comment]
@efiring are you OK if I issue a new release? 3.4.0?

**Comment by efiring** [comment]
Sure, go ahead.  Thank you.

**Pull request merged**